msprime/intervals.py

```diff
@@ -81,60 +80,66 @@ def slice(self, start=None, end=None, trim=False):  # noqa: A003
         zero rate regions such that the sequence length of the
         new rate map is end - start.
         """
-        # TODO change this to use numpy operations. We can keep the existing
-        # implentation using lists in the tests as a useful comparison.
-        positions = list(self.position)
-        rates = list(self.rate) + [0]
-
         if start is None:
             i = 0
             start = 0
         if end is None:
-            end = positions[-1]
-            j = len(positions)
+            end = self.position[-1]
+            j = len(self.position)
 
         if (
             start < 0
             or end < 0
-            or start > positions[-1]
-            or end > positions[-1]
+            or start > self.position[-1]
+            or end > self.position[-1]
             or start > end
         ):
             raise IndexError(f"Invalid subset: start={start}, end={end}")
 
         if start != 0:
-            i = bisect.bisect_left(positions, start)
-            if start < positions[i]:
+            i = np.searchsorted(self.position, start, side="left")
+            if start < self.position[i]:
                 i -= 1
-        if end != positions[-1]:
-            j = bisect.bisect_right(positions, end, lo=i)
-
-        new_positions = list(positions[i:j])
-        new_rates = list(rates[i:j])
-        new_positions[0] = start
-        if end > new_positions[-1]:
-            new_positions.append(end)
-            new_rates.append(0)
-        else:
-            new_rates[-1] = 0
+        if end != self.position[-1]:
+            j = i + np.searchsorted(self.position[i:], end, side="right")
+            if end > self.position[j - 1]:
+                j += 1
+
+        position = self.position[i:j].copy()
+        rate = self.rate[i : j - 1].copy()
+        position[0] = start
+        position[-1] = end
+
         if trim:
-            new_positions = [pos - start for pos in new_positions]
+            position -= start
         else:
-            if new_positions[0] != 0:
-                if new_rates[0] == 0:
-                    new_positions[0] = 0
+            # For the trim=False case we need to zero out all rates outside the
+            # interval [start, end], but retain the sequence_length.
+            # The simplest way to do this is to convert the position and
```

Comment:
I find it hard to believe we can't easily replicate with numpy operations. Isn't it just a straight replace of insert/append for numpy equivalents?

Reply:
Yes, that will work. But numpy's append and insert functions copy the arrays, and we're already doing so many copies around this code...

Reply:
:shrug: it's a corner case and array copies are surprisingly fast.